### DISCUSSION TITLE
Cache Modulator/Component Object

### DISCUSSION THREAD

**Discussion by dnaldoog** [RomanKubiak/ctrlr#482] [Show and tell]
### EDIT Update: 12/23/2025

In CtrlrX and Ctrlr every time you call `panel:getModulatorByName("lfofDelay")` (*for example*) or `panel:getComponent("mycontrol")` Ctrlr(X) needs to loop through every modulator on your panel looking for that modulator. If you have a lot of modulators, this is inefficient.

---
- This is the C++ code you are calling every time you use `panel:getModulatorByName():` - If your modulator is 200th in a panel, then this function will run 200 times before returning.
```c++
CtrlrModulator* CtrlrPanel::getModulator (const String& name) const
{
    // return (modulatorsByName[name]);
    for (int i=0; i<ctrlrModulators.size(); i++)
    {
        if (ctrlrModulators[i]->getProperty (Ids::name) == name)
        {
            return (ctrlrModulators[i]);
        }
    }
    return (0);
}


```
```lua
.def("getModulatorByName", &CtrlrPanel::getModulator)
```

---

It is better to cache or memoize those calls into a lua table:

- ***First call:*** `_m["lfodelay"]` → Lua detects the key doesn't exist → triggers __index → calls the expensive `panel:getModulatorByName("lfodelay")` → stores the result in the table → returns it
- ***Subsequent calls:*** `_m["lfodelay"]` → Lua finds the key in the table → returns it directly (no __index, no C++ function call)

1. The table names I use here `_m` for `getModulator()` and `_c` for `getComponent()` are entirely optional and can be changed to any legal variable/table name.

- Create a lua table `mt` in your panel and paste this code:
```lua
-- Component caching with lazy loading
mt = {}
mt.__call = function(t, k)
    return t[k]
end

mt.__index = function(t, k)
    if myDebug then
        assert(panel:getComponent(k), string.format("\n\npanel:getComponent(\"%s\") not found!", k))
    end
    
    local uiType = panel:getComponent(k):getProperty("uiType")
    
    if uiType == "uiLabel" then
        t[k] = panel:getLabelComponent(k)
    elseif uiType == "uiLCDLabel" then
        t[k] = panel:getLCDLabelComponent(k)
    elseif uiType == "uiToggleButton" then
        t[k] = panel:getToggleButtonComponent(k)
    elseif uiType == "uiButton" then
        t[k] = panel:getButtonComponent(k)
    elseif uiType == "uiImageButton" then
        t[k] = panel:getImageButtonComponent(k)
    elseif uiType == "uiCombo" then
        t[k] = panel:getComboComponent(k)
    elseif uiType == "uiListBox" then
        t[k] = panel:getListBoxComponent(k)
    elseif uiType == "uiFileListBox" then
        t[k] = panel:getFileListBoxComponent(k)
    elseif uiType == "uiImageSlider" then
        t[k] = panel:getImageSliderComponent(k)
    elseif uiType == "uiFixedSlider" then
        t[k] = panel:getFixedSliderComponent(k)
    elseif uiType == "uiFixedImageSlider" then
        t[k] = panel:getFixedImageSliderComponent(k)
    elseif uiType == "uiCustomComponent" then
        t[k] = panel:getComponent(k)
    elseif uiType == "uiProgressBar" then
        t[k] = panel:getComponent(k)
    elseif uiType == "uiTabs" then
        t[k] = panel:getComponent(k)
    else
        t[k] = panel:getModulatorByName(k)
    end
    
    return t[k]
end

_c = {}
setmetatable(_c, mt)

-- Modulator caching with lazy loading
mtm = {}
mtm.__call = function(t, k)
    return t[k]
end

mtm.__index = function(t, k)
    if myDebug then
        assert(panel:getModulator(k), sf("\n\npanel:getModulator(\"%s\") not found!", k))
    end
    
    t[k] = panel:getModulatorByName(k)
    return t[k]
end

_m = {}
setmetatable(_m, mtm)
```

----

Now just call  
`_m.lfoDelay:getModulatorValue()`
`_m("mod name with space or illegal character"):getModulatorValue()`
`_c.myComp:setProperty("componentVisibilty",0,true)`

----

[_Example_: Memoise panel](https://github.com/dnaldoog/CtrlrDemoPanels/blob/main/Panels/DEMO/DEMO%20-%20Memoize.panel)

**Comment by Ted-juh** [comment]
It's like you said, another way of doing things, but the fact that you don't have to close Ctrlr each time you add a modulator to the init file is very convenient.

Nice method :)

**Reply by dobo365** [reply]
Tbh, I don't understand why you should restart Ctrlr to get the init done...
In all my panels I have such init method (I call it AssignModulators) where I do all necessary assignments to variables
modCutoff = panel:getModulatorByName("Cutoff")

If you are changing it during development, just run it in the console window... 
AssignModulators()

May be there is something I didn't understand...

And another remark is that when writing code, be sure to be able to understand it later on for maintenance.
So avoid too cryptic things.

John is used to work with the tables and the i,v in ipairs loops and that's fine
I prefer something more readable (more old fashion / classical) even if I understand what the i,v loops are doing
What I mean by that is that you should write code the way you fell comfortable with and that will be easy to adapt later on if necessary.

**Reply by dnaldoog** [reply]
> Tbh, I don't understand why you should restart Ctrlr to get the init done... In all my panels I have such init method (I call it AssignModulators) where I do all necessary assignments to variables modCutoff = panel:getModulatorByName("Cutoff")
> 
> If you are changing it during development, just run it in the console window... AssignModulators()
> 
> May be there is something I didn't understand...
> 
> And another remark is that when writing code, be sure to be able to understand it later on for maintenance. So avoid too cryptic things.
> 
> John is used to work with the tables and the i,v in ipairs loops and that's fine I prefer something more readable (more old fashion / classical) even if I understand what the i,v loops are doing What I mean by that is that you should write code the way you fell comfortable with and that will be easy to adapt later on if necessary.

Hi Goodweather!

That is true that you can just run a function like `AssignModulators()` again to regenerate the lua variables - _All this time I didn’t know you can run a function from the console like that_!

I was more referring to the way I used to do it, which was declare the variables in the init() function, but maybe that wasn't the most efficient way looking back. In doing that I had to close the panel and restart.

The good thing about the system I have here is that you don't have to come up with new names for lua variables if the modulator name is not legal for lua. You also don't have to refer back in code to the lua variable, which you might have to look up sometimes (if it's not identical to the modulator name).

An example of why this way is more efficient in development:
Imagine you have a function that needs to access another modulator’s value. You lookup the modulator name in Ctrlr by clicking on it. With my way you can just copy the name and access with the methods above. In the case of the old way you would need to know the lua variable name that that modulator is assigned to. So in this case there’s much less work involved. 

Also in your AssignModulators function you are essentially potentially declaring dozens of global variables. With my method, variables names have restricted scope which is better programming practice it seems universally agreed.

Recently I try to name a control exactly as in the manual to keep track of everything, so for example in the Yamaha TX81-Z there is a parameter called `Operator 4-1 On/Off (bits 0-3)` - with my system I can name a modulator `_m("Operator 4-1 On/Off (bits 0-3)"):getModulatorValue()` - so for me it's more straightforward actually and involves less typing! 

Also declaring a lua variable like

`lfod= panel:getModulatorByName(“lfo delay”)` means you might have to type in` lfod:getComponent()` if you are accessing JUCE methods. More typing I guess. Also when coming back to code months later there would be no need to look up what `lfod` is actually referring to, although if you have a good naming scheme maybe that wouldn’t be necessary.

As is so often the case with coding there are a million possibilities- this is just one and may not suit everyone. 😉
